src/coreComponents/mesh/generators/HexCellBlockManager.cpp

```diff
@@ -0,0 +1,1123 @@
+  {
+    std::cout << std::setw(5) << i
+              << std::setw(5) << std::left << in[i].first 
+              << std::setw(5) << std::left << in[i].second 
+              << std::endl;
+  }
+  std::cout << std::endl
+            << std::endl;
+}
+
+/***************************************************************************************/
+/***************************************************************************************/
+
+
+/**
+ * @class MeshConnectivityBuilder
+ * @brief The MeshConnectivityBuilder to build the connectivity maps 
+ * 
+ * Initially designed for hexahedral meshes (unstructured)
+ * TODO How do we reuse this for other type of cells 
+ * Most of the code will be the same 
+ * Is templating an option? Or derivation? 
+ * 
+ * TODO and here is ONE problem: all mapping is toward Elements are not safe 
+ * since the elements may not be in the same CellBlock
+ * Check what was done - Where is this used and what for?
+ *
+ * TODO Why are storage strategies different for the mappings ?
+ * TODO Why multidimensional arrays? Isn't is more expensive? 
+ * TODO Implement specialization for regular hex mesh
+ * 
+ * TODO The storage of Faces and Edges is dependant on the Cell Types  
+ * How do we manage the CellBlocks with different types of cells?
+ * Strategy 1: allocate the space for hexahedra and keep a lot of invalid stuff 
+ *             in these vectors - not the worst idea since these meshes should be hex-dominant
+ *             do we need to store the cell type?
+ * TODO For full tetrahedral meshes we need a dedicated implementation
+ * 
+ * Options : Store if the face is a triangle or quad ? 
+ *  Store a  NO_FACE  value if  same storage space for hybrid meshes
+ *  
+ */
+class MeshConnectivityBuilder
+{
+public:
+  MeshConnectivityBuilder(CellBlockManagerBase & cellBlockManager);
+  MeshConnectivityBuilder( const MeshConnectivityBuilder & ) = delete;
+  MeshConnectivityBuilder & operator=( const MeshConnectivityBuilder & ) = delete;
+  virtual ~MeshConnectivityBuilder() = default;
+
+  localIndex numEdges() const {
+    return m_uniqueEdges.size();
+  }
+  localIndex numFaces() const {
+    return m_uniqueFaces.size();
+  }
+
+  // Dependent on Cell Type
+  virtual bool computeFaces() = 0;
+  virtual bool computeEdges() = 0;
+
+  // Independent on Cell Type
+  bool computeNodesToEdges( ArrayOfSets<localIndex> & result ) const; 
+  bool computeEdgesToNodes( array2d<localIndex>& result ) const;  
+  bool computeNodesToElements( ArrayOfArrays<localIndex>  & result ) const;
+
+  // Dependent on Cell Type
+  virtual bool computeNodesToFaces( ArrayOfSets<localIndex> & result ) const = 0 ;
+  virtual bool computeEdgesToFaces( ArrayOfSets<localIndex> & result ) const  = 0;
+  virtual bool computeFacesToNodes( ArrayOfArrays<localIndex> & result ) const = 0;
+  virtual bool computeFacesToElements( array2d<localIndex> &  result ) const  = 0;
+  virtual bool computeFacesToEdges( ArrayOfArrays<localIndex> & result ) const =0 ;
```

Comment:
I am curious: what does it change to return the arrrays ?

Reply:
Another thing, I do not understand why the storage strategies are different for the different mappings.
Do you know why? Couldit be harmonized?

Reply:
> I am curious: what does it change to return the arrrays ?

Reply:
> I am curious: what does it change to return the arrrays ?

In terms of performance, nothing. With copy elision/RVO, the returned object is constructed directly in the caller's stack frame. In terms of usage, in my opinion there are two advantages:
  * The callee doesn't have to care about previous state of the object, as it is constructed from scratch (e.g. not having to call `.resize(0)` or `.clear()` prior to populating the map.
  * The caller doesn't need a separate declaration, and can declare the immediately constructed object to be `const`, which improves readability (once I see a const object, I can remove it from a mental map of things that are changing in the function, which helps focus on those that do). In other words,
```c++
vector<int> const vals = compute();
```
is a bit cleaner than
```c++
vector<int> vals;
compute(vals);
```
It's what we already do for fundamental types like `int` and `double`, there's no reason why larger data structures have to be treated differently.

Of course, both of these are somewhat subjective and take time to get used to. There is no requirement in our coding guidelines to do it one way or the other, and there's plenty of code in GEOSX that uses out-parameter references. This is just my personal preference that I'm sharing :)

Reply:
> Another thing, I do not understand why the storage strategies are different for the different mappings.
> Do you know why? Couldit be harmonized?

The main theme is trying to use the most efficient data structure for each map. Cells in each block/subregion have a fixed shape, so cell-to-anything maps can be stored `array2d` with fixed second dimension (but this will not be the case for upcoming cell blocks representing general polyhedral grids). Faces are connected to 1 or 2 cells, so while technically it's not fixed size, we get around it by storing a dummy -1 index.

All other connectivities are unbounded, so we have to use "ragged" 2D arrays (`ArrayOfArrays` with varying second dimension. Among these, some maps remain unchanged during simulation (e.g. faces don't change shape, so `facesToEdges` and `facesToNodes` are not modified), while others do - during fracture propagation, faces and edges are getting split and duplicated, and new connectivities have to be inserted in `nodesToFaces` and `nodesToEdges`, so `ArrayOfSets` is used to allow for easy non-duplicate insertion.

It would certainly be easier if all maps were unified and just stored as `ArrayOfSets` (and it's what I do in my custom multiscale `MeshLevel` data structure), but we'd be leaving some performance on the table.